flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/collect/CollectSinkOperatorCoordinator.java

```diff
@@ -101,75 +110,81 @@ public CompletableFuture<CoordinationResponse> handleCoordinationRequest(
                 "Coordination request must be a CollectCoordinationRequest");
 
         CollectCoordinationRequest collectRequest = (CollectCoordinationRequest) request;
-        CompletableFuture<CoordinationResponse> responseFuture = new CompletableFuture<>();
-
         if (address == null) {
-            completeWithEmptyResponse(collectRequest, responseFuture);
-            return responseFuture;
+            return CompletableFuture.completedFuture(createEmptyResponse(collectRequest));
         }
 
-        executorService.submit(() -> handleRequestImpl(collectRequest, responseFuture, address));
-        return responseFuture;
+        final CompletableFuture<CoordinationResponse> responseFuture =
+                FutureUtils.supplyAsync(
+                        () -> handleRequestImpl(collectRequest, address), executorService);
+
+        ongoingRequests.add(responseFuture);
+        return responseFuture.handle(
+                (response, error) -> {
+                    if (response != null) {
+                        ongoingRequests.remove(responseFuture);
+                        return response;
+                    }
```

Comment:
Don't we always need to remove the request?
```suggestion
                    ongoingRequests.remove(responseFuture);
                    if (response != null) {
                        return response;
                    }
```